py/desiutil/annotate.py

```diff
@@ -0,0 +1,376 @@
+            if header is None:
+                header = row
+            else:
+                data.append(row)
+    log.debug(header)
+    try:
+        u = csv_unit_column(header)
+    except IndexError:
+        raise ValueError(f"{filename} does not have a unit column!")
+    try:
+        c = csv_unit_column(header, comment=True)
+    except IndexError:
+        c = None
+    for row in data:
+        log.debug("units['%s'] = '%s'", row[0], row[u])
+        units[row[0]] = row[u]
+        if c:
+            log.debug("comments['%s'] = '%s'", row[0], row[c])
+            comments[row[0]] = row[c]
+    return (units, comments)
+
+
+def yml_unit_key(data, comment=False):
+    """
+    Parameters
+    ----------
+    data : :class:`dict`
+        A dictionary resulting from a parsed YAML file.
+    comment : :class:`bool`, optional
+        If ``True``, look for a key matching 'comment'.
+
+    Returns
+    -------
+    :class:`str`
+        The key of `data` that matches.
+
+    Raises
+    ------
+    KeyError
+        If no match was found.
+    """
+    search = ('unit', )
+    if comment:
+        search = ('comment', 'description')
+    for key in data.keys():
+        for s in search:
+            if s in key.lower():
+                return key
+    raise KeyError(f"No key matching '{search[0]}' found!")
+
+
+def yml_units(filename):
+    """Parse a YAML file that contains column names and units and optionally comments.
+
+    The YAML file should contain a dictionary with a keyword like 'units' and,
+    optionally, a keyword like 'comments'.
+
+    For backwards-compatibility, the YAML file can be simply a dictionary
+    containing column names.
+
+    Parameters
+    ----------
+    filename : :class:`str` or :class:`pathlib.Path`
+        Read column definitions from `filename`.
+
+    Returns
+    -------
+    :class:`tuple`
+        A tuple containing two :class:`dict` objects for units and comments.
+        If no comments are detected, the comments :class:`dict` will be empty.
+    """
+    comments = dict()
+    # log.debug("y = yaml.safe_load('%s')", filename)
+    with open(filename, newline='') as f:
+        y = yaml.safe_load(f)
+    try:
+        u = yml_unit_key(y)
+    except KeyError:
+        log.warning(f"{filename} does not have a unit column, assuming keys are columns!")
+        u = None
+    try:
+        c = yml_unit_key(y, comment=True)
+    except KeyError:
+        c = None
+    if u:
+        units = y[u]
+    else:
+        units = y
+    if c:
+        comments = y[c]
+    return (units, comments)
+
+
+def annotate_table(table, units, comments=None, inplace=False):
+    """Add annotations to `table`.
+
+    Parameters
+    ----------
+    table : :class:`astropy.table.Table` or :class:`astropy.table.QTable`
+        A data table.
+    units : :class:`dict`
+        Mapping of table columns to units.
+    comments : :class:`dict`, optional
+        Mapping of table columns to comments.
```

Comment:
It looks like `comments` is a placeholder that doesn't to anything yet.  Let's document that, and consider even raising a `NotImplementedError` if comments is not None.

Reply:
We may want to remove that entirely, since currently there isn't an established mechanism to add comments to a `Table` in a way that could be written out to a file of any type.

Reply:
Done in [dab51a4](https://github.com/desihub/desiutil/pull/199/commits/dab51a474c40e19cc3dbbc9c2b8cfd0d4d4e199b).